### DISCUSSION TITLE
Built-in `v-model` on `<details>` and `<dialog>`

### DISCUSSION THREAD

**Discussion by antfu** [vuejs/rfcs#495] [RFC Discussions]
## Summary

Support built-in `v-model` binding for the native [`<details>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) and [`<dialog>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog).

## Links

- [Full Rendered Proposal](https://github.com/vuejs/rfcs/blob/feat/details-/active-rfcs/0000-details-element-v-model.md)
- RFC: https://github.com/vuejs/rfcs/pull/496
- Implementation PR: https://github.com/vuejs/core/pull/8048

----

# Summary

Support built-in `v-model` binding for the native [`<details>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details) and [`<dialog>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog).

# Basic example

```html
<script setup>
import { ref } from 'vue'

const show = ref(false)
</script>

<template>
  <details v-model="show">
    <summary>Summary</summary>
    <span>Details</span>
  </details>

  <dialog v-model="show">
    <form method="dialog">
      <button>OK</button>
    </form>
  </dialog>
</template>
```

When toggling the `<details>` or `<dialog>` element, the value of `show` should be reflected. Or once `show` has been modified, the details should expand/collapse automatically.

# Motivation

Currently we support `v-model` built-in on `input`, `textarea`, `select`, which is convenient to bind the value to them. However, when it comes to `<details>` and `<dialog>`, `v-model` will throw and users would need to bind manually. Which could be a bit counterintuitive.

Since `<details>` and `<dialog>` are native elements, it makes sense to have built-in support for them.

For now, users would need to manually bind them as:

```html
<script setup>
import { ref } from 'vue'

const show = ref(false)
</script>

<template>
  <details :open="show" @toggle="show = $event.target.open">
    <summary>Summary</summary>
    <span>Details</span>
  </details>

  <dialog :open="show" @close="show = false">
    <form method="dialog">
      <button>OK</button>
    </form>
  </dialog>
</template>
```

- [Playground Before](https://play.vuejs.org/#eNp9Us1qg0AQfpXtUkgLiUJbKAQjFnrroYf2uBejo7Fxf9gdE4r47p1VY0wCPS0z8/05Y8vfjAkODfA1j1xmK4PMATYmFqqSRltkLbNQsI4VVku2IOhCKKEyrRxBd/rINh7wUKS1g0ehonCQIQEqEKSpUwSqGIu2DaJWLMnqKttvBB/pd/4VvMcw9q3LsoY1a1vf7rqeGQ7UHtI3csC0qh1bawNqlBKcJdizz9r3cACFAaa2BAw8eDKKXCNlan/jr+Gl5GPjNDepit8HIxr6aggzms/SVGmty5swWa3dLEu/orN9oa1kEnCnc8IMEtN02lb8+XHx9d7fM09RehoVUTjtmi/5cLuVTE3w47Si67Z92HHgBKcFD3KC0019LfgO0bh1GDbK7Msg0zJMaBbaRmElYZVrmTwHT8HLK9k6nPcDcHK1tfrowJKj4MuZeEjNA9iVBZWDBfuv2RX2wvBqdmPqPel/6Xj3B2p5994=)
- [Playground After](https://deploy-preview-8048--vue-sfc-playground.netlify.app/#eNp1UL1OwzAQfpXDS2GA7JUbgcTGwACjF5NcUgvbZ9lOEYry7vVPGrVDp9N335/PM3tz7uU0IdszHjqvXISAcXKtsMo48hFm8DjAAoMnA7sk3QkrbEc2JOmR/uCQBY+D1AGfhOVNjUkBCUQ0TsuICQHwnylGsvDaadX9HgRb7Q95ClY0AN80jhr3MM95vSzF2VRrkZRFj1EqHeD0bKhHvYZtITxMxkj/337VmV61Li68k7Z9ryGJzKgWrcFXTUpqGu8WDeQNGIxH6hNXxRu73dx+ftzckJuy81JabAnwZvsxtpwBN2KL3g==)

# Detailed design

This is should be a compiler improvement, to be able to transform `v-model` on the `<details>` and `<dialog>` elements with `open` and `@toggle`.

# Drawbacks

It might theoretically conflict if users managed to implement a nodeTransformer on the user land to support `v-model` on `<details>` or `<dialog>`. But we are not aware of any existing library doing this.

# Alternatives

N/A

# Adoption strategy

This is a new feature and should not affect the existing code.

# Unresolved questions

N/A

**Comment by cawa-93** [comment]
How about `<dialog>` ?

**Reply by antfu** [reply]
Nice call! Added to the RFC!

**Comment by antfu** [comment]
Update: Implementation PR https://github.com/vuejs/core/pull/8048

**Comment by stafyniaksacha** [comment]
Concerning `<dialog>` element, they are two ways to open them: using [`open()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/show) or [`openModal()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLDialogElement/showModal) methods, which behave differently 

What do you think about using a `modelModifier` to support both?

**Reply by thedamon** [reply]
The native way of controlling dialog is quite strange to me (open attribute that is only for initial state and doesn't make it modal, open/openModal being separate functions instead of a modal attribute)

Abstracting that away into v-model and v-model:modal (or v-model.modal) would be really nice and avoid need of dom refs setting it up as well. The exact format of that abstraction is maybe less clear and hard to know what people would expect vs how closely to follow the native element behaviour. Should it be a modifier or a binding, should specification be required or open (vs modal) be the default?

Overall I think v-model for open and v-model.modal for openModal makes the most sense.

**Comment by skirtles-code** [comment]
As there are two states for the openness of these elements, they are similar to checkboxes from a data-binding perspective. For checkboxes, Vue supports various features beyond a basic boolean binding: e.g. `true-value`/`false-value` and binding Arrays and Sets. I can't find this mentioned in the current RFC, so I'm wondering whether these features are seen as undesirable/unnecessary for `<details>` and `<dialog>`, or are they just being left for a future iteration of the feature?

**Reply by thedamon** [reply]
I didn't realize that about checkbox models! I imagine it would make sense to handle these the same way.

**Comment by sanscheese** [comment]
I've been thinking about a potential pattern with dialogs to support async reactivity awaiting the users confirmation. In this example, `value` would only be updated when the user clicks confirm.:

``` vue
<script setup>
import { ref } from 'vue'

const show = ref(false)
const value = ref(null)
</script>

<dialog v-model="value" #default="{ internalValue }" modal :open="show">
  <form method="dialog">
    <input v-model="internalValue" />
    <button value="cancel">Cancel</button>
    <button>Confirm</button>
  </form>
</dialog>
```

This is considering the native dialog has a [returnValue](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog#handling_the_return_value_from_the_dialog)

Handling this pattern with custom components would benefit from https://github.com/vuejs/rfcs/discussions/523. Supporting the native dialog would need some sugar, and an intuitive way for the input's v-model of the temporary state for `returnValue`. Am not sure if the scope slot syntax on a native element is the answer, or am overlooking a better option.
